server/src/main/java/org/elasticsearch/index/query/functionscore/ScriptScoreQueryBuilder.java

```diff
@@ -0,0 +1,230 @@
+import org.elasticsearch.common.lucene.search.function.ScriptScoreQuery;
+import org.elasticsearch.index.query.InnerHitContextBuilder;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A query that computes a document score based on the provided script
+ */
+public class ScriptScoreQueryBuilder extends AbstractQueryBuilder<ScriptScoreQueryBuilder> {
+    public static final String NAME = "script_score";
+    public static final ParseField QUERY_FIELD = new ParseField("query");
+    public static final ParseField SCRIPT_FIELD = new ParseField("script");
+    public static final ParseField MIN_SCORE_FIELD = new ParseField("min_score");
+
+    private final QueryBuilder query;
+    private Float minScore = null;
+    private final ScriptScoreFunctionBuilder scriptScoreFunctionBuilder;
+
+
+    /**
+     * Creates a script_score query that executes the provided script function on documents that match a query.
+     *
+     * @param query the query that defines which documents the script_score query will be executed on.
+     * @param scriptScoreFunctionBuilder defines script function
+     */
+    public ScriptScoreQueryBuilder(QueryBuilder query, ScriptScoreFunctionBuilder scriptScoreFunctionBuilder) {
+        if (query == null) {
+            throw new IllegalArgumentException("script_score: query must not be null");
+        }
+        if (scriptScoreFunctionBuilder == null) {
+            throw new IllegalArgumentException("script_score: script must not be null");
+        }
+        this.query = query;
+        this.scriptScoreFunctionBuilder = scriptScoreFunctionBuilder;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public ScriptScoreQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        query = in.readNamedWriteable(QueryBuilder.class);
+        scriptScoreFunctionBuilder = in.readNamedWriteable(ScriptScoreFunctionBuilder.class);
+        minScore = in.readOptionalFloat();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(query);
+        out.writeNamedWriteable(scriptScoreFunctionBuilder);
+        out.writeOptionalFloat(minScore);
+    }
+
+    /**
+     * Returns the query builder that defines which documents the script_score query will be executed on.
+     */
+    public QueryBuilder query() {
+        return this.query;
+    }
+
+    /**
+     * Returns the script function builder
+     */
+    public ScriptScoreFunctionBuilder scriptScoreFunctionBuilder() {
+        return this.scriptScoreFunctionBuilder;
+    }
+
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        if (query != null) {
```

Comment:
Is this check necessary? We have a null check for query in the constructor so I don't think this can be null here?